### PR TITLE
Fix(Android): Remove Filepicker and use SAF instead.

### PR DESCRIPTION
**Pull request by hank121314**
## Summary

This PR intent to remove `com.nononsenseapps:filepicker` and use Storage Access Framework (SAF) instead. 
With this change users using Android 11 will be available to select the file/directory from external storage.

### For Directory

- Using `ACTION_OPEN_DOCUMENT_TREE` to get the content uri.
- Create `DocumentFile` with `DocumentFile.fromTreeUri`.
- Get all files in directory with `DocumentFile.listFiles`.
- Open the file and get the filedescripor with `context.getContentResolver().openFileDescriptor`.
- To get real filepath, using `Files.readSymbolicLink(filedescripor)`.

### For File

- Using `ACTION_OPEN_DOCUMENT` to get the content uri.
- Create `DocumentFile` with `DocumentFile.fromSingleUri`.
- Open the file and get the filedescripor with `context.getContentResolver().openFileDescriptor`.
- To get real filepath, using `Files.readSymbolicLink(filedescripor)`.

## Test

I have tested with the apk below in android 11. 
- Release APK: [citra-android-11-sdcard-fix](https://drive.google.com/file/d/19JgE3sjxbJddHinRmFIoJKvN2JCnbK96/view?usp=sharing)

## Thanks

This is my first contribution, please feel free to critique, and thanks for your code review 😄 .

### CODE REVIEW

**Comment by SachinVin** [comment]
Hi, I'm curious if cpp side really works without any modification, also you may want to disable `requestLegacyExternalStorage` in the [AndroidManifest.xml](https://github.com/citra-emu/citra-android/blob/master/src/android/app/src/main/AndroidManifest.xml#L32), or bump up the targetSdk so it doesn't interfere with testing.

**Comment by xperia64** [comment]
While this does work for local CIAs and game folders, it does not handle Citra's data folder which would be required before removing those permissions and bumping the SDK version.

I have mixed feelings completely relying on `/proc/self/fd` in the way it is here.

When opening a CIA from a non-local source such as Google Drive, it handles failure in a file-not-found way. I have not yet tried an alternate DocumentTree provider to see how selecting a game folder there that would fair (e.g. Samba or something; Apparently RetroArch can provide DocumentTrees too?)

**Comment by hank121314** [comment]
Hi @SachinVin! Thanks for the review!

At first, this RP will resolve the sdcard issue only when targetSdk is lower than 30.
But now, I think we should remove all system permissions like `WRITE_EXTERNAL_STORAGE` and `WRITE_MEDIA_STORAGE` since they will no longer provide any additional access.

**Comment by xperia64** [comment]
> Hi @SachinVin! Thanks for the review!
> 
> At first, this RP will resolve the sdcard issue only when targetSdk is lower than 30.
> But now, I think this PR should remove all system permissions like `WRITE_EXTERNAL_STORAGE` and `WRITE_MEDIA_STORAGE` since they will no longer provide any additional access.

No, WRITE_EXTERNAL_STORAGE is still required for Citra's data folder located at /sdcard/citra-emu, which contains the emulated SD card that contains all installed CIA content. I already removed permissions as a test just to make sure, and it still needs them.

**Comment by hank121314** [comment]
> No, WRITE_EXTERNAL_STORAGE is still required for Citra's data folder located at /sdcard/citra-emu, which contains the emulated SD card that contains all installed CIA content. I already removed permissions as a test just to make sure, and it still needs them.

Yes! Although using `ACTION_OPEN_DOCUMENT_TREE` to grant the write permission, we still can not access it with `fopen`.
But AFAIK, after targeting SDK version over 30, `WRITE_EXTERNAL_STORAGE` will no longer provide any additional access.
The older File API can again be used but "only" when accessing the public "shared storage" folders (DCIM, Music, etc.), or your app "private" directory. 

Thanks for the review! 😄

**Comment by RonzelIGuess** [comment]
Screenshot_20210811-085640
My roms have these weird cracked icon plus if i open any games i just crashes, im using android 9

**Comment by DatEmulationGuy** [comment]
It took me like 5 minutes to install Pokemon X with this test version. The official version from the store and famous fork also take long to install whole games on Android 11, I'm sure it was like half the time on Android 10 and on a weaker phone :( The screen also turns black while it's installing but if I leave it will successfully install the game. Can you add a better installation interface with percentage or something

**Comment by AbandonedCart** [comment]
@SachinVin https://developer.android.com/about/versions/11/privacy/storage#test-scoped-storage provides methods to test scoped storage without having to modify the manifest. This can help to avoid forgetting to restore the flag and subjecting anyone to the broken version of scoped storage attempted in Android 10.

@hank121314 

> If your app targets Android 11, both the WRITE_EXTERNAL_STORAGE permission and the WRITE_MEDIA_STORAGE privileged permission no longer provide any additional access. 

> In order to access media files using native files methods on Android 10, you must also request the READ_EXTERNAL_STORAGE permission

No additional information is listed for Android 11, other than fopen() is still available. The current AndroidManifest does not include READ_EXTERNAL_STORAGE.

I have not researched the full process for it, but you may also be able to use the exemption for document management apps that enables use of the MANAGE_EXTERNAL_STORAGE permission. It applies to apps that need access to specific files outside of the app-specific directories and shared storage that SAF is not a sufficient solution.
https://support.google.com/googleplay/android-developer/answer/10467955

TL;DR SAF is a PITA, but best of luck with it.

**Comment by bravebg** [comment]
Is this the last version of the official citra build or is it based off of some older version ? Wonder when super mario land 3d will be playing;(

**Comment by hank121314** [comment]
Hi @SachinVin!
I have made some changes to this PR.
Could you help me to review it? Thanks!

**Comment by hank121314** [comment]
clang-format on my locale might break some codes. I will fix it later!

**Comment by SachinVin** [comment]
H!, I'll let xperia64 review, given they are much more experienced with this.

**Comment by xperia64** [comment]
I'm afraid I don't agree with Dolphin's approach of attempting to parse DocumentsContract-related URIs and extracting/constructing paths manually from them.

Google pretty much says "don't do that" here: https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_DOCUMENT_ID

```Unique ID of a document. This ID is both provided by and interpreted by a DocumentsProvider, and should be treated as an opaque value by client applications```

In another section of documentation, they note explicitly to not parse this ID, which to me is pretty clear that they don't want you messing with parsing/building URIs yourself by guessing the format (Google has been known to make good on most of their threats of things they tell you not to do, so if they ended up obfuscating Document URIs, it really wouldn't surprise me)

My original intent for implementing SAF in Citra was to thoroughly abstract away filesystem traversal and file opening into a "Path" class which would provide Android and normal filesystem implementations. The class would include a "root" (on Android, the URI of a top-level Document tree), and a normal subpath string, with JNI calls to query the DocumentsProvider as needed.

However, the main issue with any approach at the moment is it's not very maintainable given how divergent upstream citra and citra-android are. The priority right now really should be attempting to get the Android port merged into upstream.

**Comment by hank121314** [comment]
Hi @xperia64! Thanks for the review once again! 😃 
As for DocumentsContract-related URIs, I refer to [DocumentsContractApi21.java](https://android.googlesource.com/platform/frameworks/support/+/5fdfbc2/v4/api21/android/support/v4/provider/DocumentsContractApi21.java) and [DocumentsContractApi19](https://android.googlesource.com/platform/frameworks/support/+/5fdfbc2/v4/kitkat/android/support/v4/provider/DocumentsContractApi19.java).
Although Google says don't do that. They are still using this in their source code.
And I also strongly agree with do not extract/construct content URI manually.
But, I have also try to use a top-level `DocuemntFile` and a normal subpath string to traverse the path of content URI.
It results in really poor performance on any I/O operations.
Currently, I think the only way to have an efficient I/O is to extract/construct content URI manually.

> so if they ended up obfuscating Document URIs, it really wouldn't surprise me

I believe that the day will come 🤣.
When the day comes I will try my best to fix it asap.

**Comment by xperia64** [comment]
> 
> As for DocumentsContract-related URIs, I refer to [DocumentsContractApi21.java](https://android.googlesource.com/platform/frameworks/support/+/5fdfbc2/v4/api21/android/support/v4/provider/DocumentsContractApi21.java) and [DocumentsContractApi19](https://android.googlesource.com/platform/frameworks/support/+/5fdfbc2/v4/kitkat/android/support/v4/provider/DocumentsContractApi19.java).
> Although Google says don't do that. They are still using this in their source code.

Unless I'm missing something, I'm failing to see where in those files Google is manually parsing or constructing URIs. Using DocumentsContract URI opaquely as intended is fine, but the point is, _don't parse or build your own URIs_

> > so if they ended up obfuscating Document URIs, it really wouldn't surprise me
> 
> I believe that the day will come 🤣.
> When the day comes I will try my best to fix it asap.

I would rather do it correctly the first time.

**Comment by hank121314** [comment]
> Unless I'm missing something, I'm failing to see where in those files Google is manually parsing or constructing URIs. Using DocumentsContract URI opaquely as intended is fine, but the point is, _don't parse or build your own URIs_

Yeah! 
Sorry for misunderstood the meaning.
I delete the comments above.

But is there any way to have efficient I/O operations?
Since using `DocumentFile` to traverse the path of content URI will get terrible performance 😢.

**Comment by hehk1234** [comment]
Hi, can you provide this fix on citra mmj?

**Comment by Gamemulatorer** [comment]
Happy Halloween 🎃 everyone!

What is the status of this pr?

**Comment by Gamer64ytb** [comment]
Hi, when I compile your code I get this error:

C/C++ release|arm64-v8a : CMake Error: CMake can not determine linker language for target: boost_iostreams

**Comment by hank121314** [comment]
> Hi, when I compile your code I get this error:
> 
> C/C++ release|arm64-v8a : CMake Error: CMake can not determine linker language for target: boost_iostreams

Hi, @Gamer64ytb.

This PR uses `boost::iostreams` to open `fstream` from the file descriptor.  
So you will have to add `boost_iostreams` to external/boost manually.
Or maybe set `USE_SYSTEM_BOOST` in CMake to `on`.

After this PR is ready to merge, I will create another PR in [citra-emu/ext-boost](https://github.com/citra-emu/ext-boost/tree/master).

**Comment by hank121314** [comment]
// @xperia64  friendly ping 😃

Would you be willing to review this? Thanks!

**Comment by xperia64** [comment]
> // @xperia64 friendly ping smiley
> 
> Would you be willing to review this? Thanks!

Yep sorry, reviewing this is still queued up.


I'm waiting on me completing my own SAF implementation and integration for DraStic so I don't accidentally steal things from this. Sorry for the conflict of interest, just trying to avoid any problems.

**Comment by hank121314** [comment]
> I'm waiting on me completing my own SAF implementation and integration for DraStic so I don't accidentally steal things from this. Sorry for the conflict of interest, just trying to avoid any problems.

Thanks! Just want to make sure this PR will not be forgotten.

It will be my pleasure if some of my code will help you to develop your own SAF implementation!
And thanks for your code review once again! 😄

**Comment by TheXpertXD** [comment]
Is this still in the works?

**Comment by Gamemulatorer** [comment]
> Some

Out of topic DraStic will get updated this year or never?

**Comment by JosJuice** [comment]
So... the reviewing of this PR has been stalled for a while now. As far as I know this is because xperia64 still has to finish the SAF work for DraStic, though it has been a few months since I last checked, so I could be wrong. In the interest of hopefully advancing this PR, I'd like to offer my comments, in case they're useful to others in the Citra project. I'm not familiar enough with Citra's needs that I would be able to give this an LGTM on my own, but I have experience with SAF from doing the SAF migration for Dolphin.

The big problem of migrating to SAF is that SAF insists that you should use "content URIs" for addressing files, yet Citra and Dolphin both have lots of existing C++ code that assumes that files can be addressed using paths that are delimited by slashes and can be stored in `std::string`s. Updating all the C++ code to use some kind of abstracted class that can handle both paths and content URIs would be a lot of work (though it wouldn't be impossible – I believe xperia64 started an attempt at it before this PR was opened?), so what both this PR and my Dolphin PR does is to try to invent a way to squeeze a content URI into an std::string in a way that can be reversed once it's time to open the file.

Squeezing a content URI into a string isn't too hard – just take the bytes of the URI and put them in the string. Or well, this is Java, so take the UTF-16 code points and convert them to UTF-8, but that's not what's important here. What's important is the other detail I mentioned: That existing C++ code assumes that path components are delimited by slashes and that it can add and remove path components. This is where this PR and my Dolphin PR really start to differ.

What I did for Dolphin was to copy the URI into a string, and if C++ then wanted to add a path component, just... let it add a slash and filename, and then once it's time to open the file, try to "unmangle" the strange URI/path mix we now have. It works, but it's ugly, xperia64 considers it too unreliable (which I can understand), and the already terrible performance of SAF becomes abysmal when you have deeply nested paths and unmangling them becomes an O(number of path components) operation **for each file**.

What this PR does instead is to have two ways of squeezing URIs into strings. The first approach is to just copy the URI into a string – like Dolphin, but without the "unmangling". When using this approach, the existing C++ code isn't able to do anything with path components. The second approach is that it "pretends" that the user directory is called `/`, and then whenever some C++ code opens a file using a path that starts with `/`, it gets turned into a URI in the user directory.

What's clever about the `/` scheme is that to the C++ code, it really looks just like a regular path. All operations that inspect paths and add or remove parts of paths will simply work – as long as the code then goes through the right wrapper functions once it's time to do something like open the file or enumerate the contents of a directory, but that is generally already the case. Also, the PR implements caching for SAF lookups for this `/` scheme, mitigating the performance problems I mentioned having in Dolphin. SAF is still slow, but it's a lot less slower this way.

The big advantage that this gives Citra is that letting the user select a location for the user directory using SAF actually is feasible. Indeed, this PR implements it. This is something I unfortunately haven't been able to implement in Dolphin using the strategy I chose, in part because a good bunch of extra C++ code would have to be changed to play nice with URIs (or mangled URIs), but in much greater part because the performance would've been complete garbage. Really, I cannot overstate this. SAF is so slow, and seeing caching implemented in this PR is promising.

(For those of you less familiar with Android: If you don't let the user select a location using SAF, the only location the app can use is the app-specific directory inside Android/data/. Google made that directory a lot harder to access starting in Android 11. What they wanted to do was to prevent apps from accessing the data of other apps, but, well, when the user is using a file manager they're using an app, and the new rule applies to file managers in the same way as any other app...)

But there is a downside to this PR's strategy too. The `/` scheme is only for the user directory. The other approach of just stuffing the URI inside the string is fine for single files, but as the code is right now, as far as I can tell, existing C++ code simply cannot handle folders outside of the user directory correctly. I don't know if this is a big deal for Citra or not. It seems like the game scanning is implemented in Java, and this PR rewrote that code to use URIs, which solves the problem for the game list. Maybe someone who knows Citra better than me can answer: Are there are features or use cases where C++ needs to be able to deal with folders that are outside of the user directory? (Enumerating files in a folder, creating or deleting files, and so on.)

And finally, a note about the UX in this PR. The first thing that happens when you launch Citra is that you see the following message:

> This software will run games for the Nintendo 3DS handheld game console. No game titles are included.
>
> Before you run, please place your rightfully owned 3DS game files onto your device storage.

In master, this is followed by a fairly self-explanatory prompt to allow Citra access to storage. In this PR, it's instead followed by a SAF folder picker. Because this SAF folder picker comes immediately after a text talking about 3DS game files, and because there's no explanation of what the SAF folder picker is for (thanks Google for not letting us set a custom string to be shown inside the folder picker activity), it's very easy to assume that the folder picker is for selecting your games, even though it's actually for selecting where you want to put the user directory. For some subset of users, I'm sure this will cause a mess where the user directory is a directory that contains games, or is just some random strange directory if the user had no idea what folder to pick. This UX problem is something I really think needs to be addressed before the PR can be merged.

Just as an idea: I would guess that there's a good number of users that don't want to have to think about folders and would pick a poor location if the app forces them to select where the user directory should be. How about having the user directory in Android/data/ by default, and the first time the app is launched, showing a dialog where you can either press a button to select a custom location or just dismiss the dialog if you're fine with the default location? Or perhaps even having Android/data/ as the default and not having a prompt at first launch... but then you'd better have a way to move the contents of the user folder when selecting a new location, so that the contents of people's user directories don't become "stuck" inside Android/data/.

If not, then I suppose simply some sort of explanation before we show the folder picker.

**Comment by bunnei** [comment]
@hank121314 apologies for the great delay on this -- there was a big gap in Citra Android development, coupled with xperia's concerns above taking a little longer than expected to sort out. At this point, Android development has been merged into Citra's main repository (https://github.com/citra-emu/citra/). Would you be interested in rebasing & opening this PR there, or would you prefer another developer take this over?

(And either way, thank you for this great work thus far)

**Comment by hank121314** [comment]
> At this point, Android development has been merged into Citra's main repository ([citra-emu/citra](https://github.com/citra-emu/citra/)).

Thanks for the reminder!

> Would you be interested in rebasing & opening this PR there, or would you prefer another developer take this over?

I will open a PR in the main repository when I complete the rebasing!

**Comment by bunnei** [comment]
@hank121314 Great to hear! Again, apologies this took so long. I appreciate the work that you put into this and would love to see it merged into our main branch. I'll make sure this does not fall through the cracks this time. Thanks again!

**Comment by hank121314** [comment]
> @hank121314 Great to hear! Again, apologies this took so long. I appreciate the work that you put into this and would love to see it merged into our main branch. I'll make sure this does not fall through the cracks this time. Thanks again!

There's no need to apologize! I understand there are lots of work in citra team. 
And many thanks for building this awesome emulator ❤️ !

**Comment by bunnei** [comment]
@hank121314 Gentle ping here, since I have not seen or heard anything for a few weeks 🙂

Is this something that you still plan to work on? It would be great if you could rebase this change and PR it to https://github.com/citra-emu/citra, as all Android development has moved to this repository.

Also, if you are in our [Discord server](https://discord.gg/FAXfZV9), feel free to reach out to me bunnei#0166 (DM is best) if it is easier for you to discuss there.

Thank you!

**Comment by hank121314** [comment]
Hi, @bunnei many thanks for keep tracking on this ❤️ .

> Is this something that you still plan to work on?

I will created a WIP PR in [citra-emu/citra](https://github.com/citra-emu/citra) after my SAF initial implementation(in this weekend).
I also review some code in this PR and doing some refactor.

> Also, if you are in our [Discord server](https://discord.gg/FAXfZV9), feel free to reach out to me bunnei#0166 (DM is best) if it is easier for you to discuss there.

Got it! Many thanks for the support!

**Comment by bunnei** [comment]
@hank121314 Great to hear! Apologies for my impatience, I'm just excited to see SAF come to Citra 😄 

Looking forward to it and thank you again!
